Add release notes for Cookstyle 5.4 (#266)
chef/cookstyle: RELEASE_NOTES.md

#### Chef/UseInlineResourcesDefined

> The `UseInlineResourcesDefined` checks for resources that call the `use_inline_resources` method. Chef Infra Client 13 made the inline resources mode the default for all resources, so this call is no longer necessary.

Reviewer: ```suggestion
The `UseInlineResourcesDefined` cop checks for resources that call the `use_inline_resources` method. Chef Infra Client 13 made the inline resources mode as the default for all resources, so this call is no longer necessary.
```

`Enabled by default`: True

`Autocorrects`: Yes

#### Chef/IncludingOhaiDefaultRecipe

> The `IncludingOhaiDefaultRecipe` detects cookbooks that include the `ohai::default` recipe. The ohai cookbook's default recipe was used to install custom Ohai plugins located in the cookbook's `files` directory. This required forking the ohai community cookbook to add your own plugins, which we no longer recommend doing. The Ohai cookbook provides an [ohai_plugin resource](https://github.com/chef-cookbooks/ohai/#ohai_plugin) which can be used to install plugins and Chef Infra Client. Chef Infra Client. Additionally plugins can be placed in a `/ohai` directory in any cookbook to have Chef Infra Client install install these at runtime. Note: This feature requires Chef Infra *Server* 12.18 or later.

Reviewer: ```suggestion
The `IncludingOhaiDefaultRecipe` cop detects cookbooks that include the `ohai::default` recipe. The ohai cookbook's default recipe was used to install custom Ohai plugins located in the cookbook's `files` directory. This required forking the ohai community cookbook to add your own plugins, which we no longer recommend doing. The Ohai cookbook provides an [ohai_plugin resource](https://github.com/chef-cookbooks/ohai/#ohai_plugin), which can be used to install plugins and Chef Infra Client. Additionally, plugins can be placed in a `/ohai` directory in any cookbook to have Chef Infra Client install these at runtime. Note: This feature requires Chef Infra *Server* 12.18 or later.
```

> The `IncludingOhaiDefaultRecipe` detects cookbooks that include the `ohai::default` recipe. The ohai cookbook's default recipe was used to install custom Ohai plugins located in the cookbook's `files` directory. This required forking the ohai community cookbook to add your own plugins, which we no longer recommend doing. The Ohai cookbook provides an [ohai_plugin resource](https://github.com/chef-cookbooks/ohai/#ohai_plugin) which can be used to install plugins and Chef Infra Client. Chef Infra Client. Additionally plugins can be placed in a `/ohai` directory in any cookbook to have Chef Infra Client install install these at runtime. Note: This feature requires Chef Infra *Server* 12.18 or later.

Reviewer: One more thing - please clarify what "This" is per the context of "This required forking the ohai community cookbook" Thanks!

Author: Fixed

> The `IncludingOhaiDefaultRecipe` detects cookbooks that include the `ohai::default` recipe. The ohai cookbook's default recipe was used to install custom Ohai plugins located in the cookbook's `files` directory. This required forking the ohai community cookbook to add your own plugins, which we no longer recommend doing. The Ohai cookbook provides an [ohai_plugin resource](https://github.com/chef-cookbooks/ohai/#ohai_plugin) which can be used to install plugins and Chef Infra Client. Chef Infra Client. Additionally plugins can be placed in a `/ohai` directory in any cookbook to have Chef Infra Client install install these at runtime. Note: This feature requires Chef Infra *Server* 12.18 or later.

Author: Fixed

`Enabled by default`: True

`Autocorrects`: No

#### Chef/IncludingXMLRubyRecipe

> The `IncludingXMLRubyRecipe` detects cookbooks that include the `xml::ruby` recipe. This recipe was used to install the `nokogiri` gem into Chef Infra Client for use in parsing XML in cookbooks. Nokogori ships with Chef Infra Client 13+, so this recipe is no longer necessary.

Reviewer: ```suggestion
The `IncludingXMLRubyRecipe` cop detects cookbooks that include the `xml::ruby` recipe. This recipe was used to install the `nokogiri` gem into Chef Infra Client for use in parsing XML in cookbooks. Nokogori ships with Chef Infra Client 13+, so this recipe is no longer necessary.
```

> The `IncludingXMLRubyRecipe` detects cookbooks that include the `xml::ruby` recipe. This recipe was used to install the `nokogiri` gem into Chef Infra Client for use in parsing XML in cookbooks. Nokogori ships with Chef Infra Client 13+, so this recipe is no longer necessary.

Author: Fixed

`Enabled by default`: True

`Autocorrects`: Yes

#### Chef/LegacyYumCookbookRecipes

The `LegacyYumCookbookRecipes` cop detects usage of legacy [yum](https://supermarket.chef.io/cookbooks/yum) cookbook recipes. These recipes were removed from the yum cookbook in verson 3.0, which was released in 2013. Cookbooks should be updated to use the new individual yum cookbooks that contain the same functionality.

`Examples`

Legacy recipes:
```ruby
  include_recipe 'yum::epel'
  include_recipe 'yum::ius'
```

New cookbooks containing the same functionality:
```ruby
  include_recipe 'yum-epel'
  include_recipe 'yum-ius'
```

`Enabled by default`: True

`Autocorrects`: No

#### Chef/DefaultMetadataMaintainer

The `DefaultMetadataMaintainer` cop detects the default `maintainer` and `maintainer_email` fields generated by the `chef generate cookbook` command. Cookbook metadata should be updated with actual maintainer information.

`Enabled by default`: True

`Autocorrects`: No

#### Chef/UsesDeprecatedMixins

> The `UsesDeprecatedMixins` cop detects various legacy mixins used in `HWRP` and `LWRP` resources. These mixins are no longer necessary and were removed in Chef Infra Client 14, which will cause errors running these resources on modern Chef Infra Client releases.

Reviewer: ```suggestion
The `UsesDeprecatedMixins` cop detects various legacy mixins used in `HWRP` and `LWRP` resources. These mixins are no longer necessary and were removed in Chef Infra Client 14, which will cause errors when running these resources on modern Chef Infra Client releases.
```

> The `UsesDeprecatedMixins` cop detects various legacy mixins used in `HWRP` and `LWRP` resources. These mixins are no longer necessary and were removed in Chef Infra Client 14, which will cause errors running these resources on modern Chef Infra Client releases.

Author: Fixed

`Enabled by default`: True

`Autocorrects`: Yes

#### Chef/MinitestHandlerUsage

> The `MinitestHandlerUsage` cop detects cookbooks that depend on the [minitest-handler](https://supermarket.chef.io/cookbooks/minitest-handler) cookbook. This cookbook provides a handler for running [minitests](https://github.com/seattlerb/minitest) after each Chef Infra Client run. The handler is no longer maintained and we highly recommend using InSpec for validating system configuration post Chef Infra Client run. InSpec includes over 100 build in resources and integrates with Automate to report the results of system state validation.

Reviewer 2: ```suggestion
The `MinitestHandlerUsage` cop detects cookbooks that depend on the [minitest-handler](https://supermarket.chef.io/cookbooks/minitest-handler) cookbook. This cookbook provides a handler for running [minitests](https://github.com/seattlerb/minitest) after each Chef Infra Client run. The handler is no longer maintained and we highly recommend using InSpec for validating a system configuration after a Chef Infra Client run. InSpec includes over 100 build in resources and integrates with Automate to report the results of a system state validation.
```

> The `MinitestHandlerUsage` cop detects cookbooks that depend on the [minitest-handler](https://supermarket.chef.io/cookbooks/minitest-handler) cookbook. This cookbook provides a handler for running [minitests](https://github.com/seattlerb/minitest) after each Chef Infra Client run. The handler is no longer maintained and we highly recommend using InSpec for validating system configuration post Chef Infra Client run. InSpec includes over 100 build in resources and integrates with Automate to report the results of system state validation.

Reviewer: ```suggestion
The `MinitestHandlerUsage` cop detects cookbooks that depend on the [minitest-handler](https://supermarket.chef.io/cookbooks/minitest-handler) cookbook. This cookbook provides a handler for running [minitests](https://github.com/seattlerb/minitest) after each Chef Infra Client run. The handler is no longer maintained and we highly recommend using InSpec for validating system configuration post-Chef Infra Client run. InSpec includes over 100 builds in resources and integrates with Automate to report the results of system state validation.
```

Reviewer: Please check if the phrasing of "100 builds in resources" in the last sentence is the intended phrasing.

Author: 🤦‍♂ nope

Author: Fixed

`Enabled by default`: True

`Autocorrects`: No

#### Chef/WindowsVersionHelper

> The `WindowsVersionHelper` cop detects the usage of the various Windows version helpers that shipped in the [windows](https://supermarket.chef.io/cookbooks/windows) cookbook. Nearly all of the resources that originally shipped in Windows cookbook are now part of Chef Infra Client itself. We'd like to help move users away from this cookbook so they can avoid the added dependency in their environment. Instead of using the Windows version helpers we suggest using data from Ohai itself to determine what version of Windows is running on a node.

Reviewer 2: ```suggestion
The `WindowsVersionHelper` cop detects the usage of the various Windows version helpers that shipped in the [Windows](https://supermarket.chef.io/cookbooks/windows) cookbook. Nearly all of the resources that originally shipped in the Windows cookbook are now part of Chef Infra Client itself. We'd like to help move users away from this cookbook so they can avoid the added dependency in their environment. Instead of using the Windows version helpers we suggest using data from Ohai itself to determine what version of Windows is running on a node.
```

Author: Fixed

> The `WindowsVersionHelper` cop detects the usage of the various Windows version helpers that shipped in the [windows](https://supermarket.chef.io/cookbooks/windows) cookbook. Nearly all of the resources that originally shipped in Windows cookbook are now part of Chef Infra Client itself. We'd like to help move users away from this cookbook so they can avoid the added dependency in their environment. Instead of using the Windows version helpers we suggest using data from Ohai itself to determine what version of Windows is running on a node.

Reviewer: ```suggestion
The `WindowsVersionHelper` cop detects the usage of the various Windows version helpers that shipped in the [windows](https://supermarket.chef.io/cookbooks/windows) cookbook. Nearly all of the resources that originally shipped in Windows cookbook are now part of Chef Infra Client itself. We would like to help move users away from this Windows cookbook so they can avoid the added dependency in their environment. Instead of using the Windows version helpers, we suggest using data from Ohai itself to determine what version of Windows is running on a node.
```

`Examples`

A cookbook using the Windows version helper:
```ruby
  Windows::VersionHelper.nt_version == 6.1
```

The same version comparison could be done using Ohai attributes
```ruby
  node['platform_version'].to_f == 6.1
```

`Enabled by default`: True

`Autocorrects`: No

#### Chef/WindowsZipfileUsage

> The `WindowsZipfileUsage` detects the usage of the `windows_zipfile` resource in the [windows](https://supermarket.chef.io/cookbooks/windows) cookbook. This is one of the remaining resources in the windows cookbook that is not present in Chef Infra Client itself. In Chef Infra Client 15.0 we introduced the [archive_file](https://docs.chef.io/resource_archive_file.html) resource, which is a cross platform resource for uncompressing archives. This resource uses an ulta-fast compression library with support for multiple compression formats, and should be used instead of the legacy `windows_zipfile` resource.

Reviewer 2: ```suggestion
The `WindowsZipfileUsage` detects the usage of the `windows_zipfile` resource in the [Windows](https://supermarket.chef.io/cookbooks/windows) cookbook. This is one of the remaining resources in the windows cookbook that is not present in Chef Infra Client itself. In Chef Infra Client 15.0 we introduced the [archive_file](https://docs.chef.io/resource_archive_file.html) resource, which is a cross platform resource for uncompressing archives. This resource uses an ulta-fast compression library with support for multiple compression formats, and should be used instead of the legacy `windows_zipfile` resource.
```

Author: Fixed

> The `WindowsZipfileUsage` detects the usage of the `windows_zipfile` resource in the [windows](https://supermarket.chef.io/cookbooks/windows) cookbook. This is one of the remaining resources in the windows cookbook that is not present in Chef Infra Client itself. In Chef Infra Client 15.0 we introduced the [archive_file](https://docs.chef.io/resource_archive_file.html) resource, which is a cross platform resource for uncompressing archives. This resource uses an ulta-fast compression library with support for multiple compression formats, and should be used instead of the legacy `windows_zipfile` resource.

Reviewer: ```suggestion
The `WindowsZipfileUsage` detects the usage of the `windows_zipfile` resource in the [windows](https://supermarket.chef.io/cookbooks/windows) cookbook. The `windows_zipfile` resource is one of the remaining resources in the windows cookbook that is not present in Chef Infra Client itself. In Chef Infra Client 15.0, we introduced the [archive_file](https://docs.chef.io/resource_archive_file.html) resource, which is a cross-platform resource for uncompressing archives. This new resource uses an ultra-fast compression library with support for multiple compression formats, and should be used instead of the legacy `windows_zipfile` resource.
```

`Enabled by default`: True

`Autocorrects`: No

#### Chef/NodeMethodsInsteadofAttributes

> The `NodeMethodsInsteadofAttributes` cop detects cookbooks that reference node attributes in the legacy `node.some.value` format. Chef Infra Client 13 officially deprecated this attribute format. This cop detects the usage of common Ohai attributes and can autocorrect those, but it won't detect cookbook attributes, which will also need to be updated.

Reviewer 2: ```suggestion
The `NodeMethodsInsteadofAttributes` cop detects cookbooks that reference node attributes in the legacy `node.some.value` format. Chef Infra Client 13 officially deprecated this attribute format. This cop detects the usage of common Ohai attributes and can autocorrect those. But it won't detect cookbook attributes which also need to be updated.
```

`Enabled by default`: True

`Autocorrects`: Yes

`Examples`

Node attributes in the method format:
```ruby
  node.platform
  node.platform_version
  node.my_cookbook.attribute1
```

The preferred format:
```ruby
  node['platform']
  node['platform_version']
  node['my_cookbook']['attribute1']
```

#### Chef/IncludingMixinShelloutInResources

> The `IncludingMixinShelloutInResources` cop detects resources that require and include the Chef Shellout Mixin helper. Chef Infra Client now includes this mixin helper automatically for all resources so this require and include can be removed.

Reviewer: Are there other words supposed to be used instead of "require and include in the last sentence? ("this require and include can be removed."

Author: require and include are the actual Ruby calls that are being called

> The `IncludingMixinShelloutInResources` cop detects resources that require and include the Chef Shellout Mixin helper. Chef Infra Client now includes this mixin helper automatically for all resources so this require and include can be removed.

Reviewer 2: ```suggestion
The `IncludingMixinShelloutInResources` cop detects resources that require and include the Chef Shellout Mixin helper. Chef Infra Client now includes the Chef Shellout Mixin helper automatically for all resources, so require and include statements can be removed from all resources.
```
